Write support

thalesfragoso opened the pull request:
I started to work on this write support, right now it's very rough on the edges, it also needs more testing. I did a very simple test under linux (see `write_test.rs`) and it worked as expected.

I still need to add some open support and really deal with modes and file attributes, but I would like some feedback before making more changes.

thalesfragoso commented on src/lib.rs:
```diff
@@ -541,7 +764,7 @@ mod tests {
             _reason: &str,
         ) -> Result<(), Self::Error> {
             // Actual blocks taken from an SD card, except I've changed the start and length of partition 0.
```
I added a fake `info_sector`, because I didn't want to change all the blocks, if you still have the SD image you can copy in the third block and run cargo test, I think it should be fine, I tested this on other fat32 images.

thalesfragoso commented:
It might need some more testing, but everything seems to be working fine. I still have to update the `README` and maybe add some more output to the examples, if everything is fine I can do that so you can merge.
CC:
@thejpster

Wassasin commented:
With an empty pre-existing `readme.txt` I get:
```
thread 'main' panicked at 'attempt to subtract with overflow', src/fat.rs:1037:24
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.37/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.37/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:76
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:60
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1030
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1412
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:64
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:196
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:210
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:473
  11: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:380
  12: rust_begin_unwind
             at src/libstd/panicking.rs:307
  13: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
  14: core::panicking::panic
             at src/libcore/panicking.rs:49
  15: embedded_sdmmc::fat::Fat32Volume::cluster_to_block
             at src/fat.rs:1037
  16: embedded_sdmmc::Controller<D,T>::find_data_on_disk
             at ./src/lib.rs:686
  17: embedded_sdmmc::Controller<D,T>::write
             at ./src/lib.rs:583
  18: write_test::main
             at examples/write_test.rs:160
  19: std::rt::lang_start::{{closure}}
             at /rustc/6ef275e6c3cb1384ec78128eceeb4963ff788dca/src/libstd/rt.rs:64
  20: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:49
  21: std::panicking::try::do_call
             at src/libstd/panicking.rs:292
  22: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:80
  23: std::panicking::try
             at src/libstd/panicking.rs:271
  24: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  25: std::rt::lang_start_internal
             at src/libstd/rt.rs:48
  26: std::rt::lang_start
             at /rustc/6ef275e6c3cb1384ec78128eceeb4963ff788dca/src/libstd/rt.rs:64
  27: main
  28: __libc_start_main
  29: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Wassasin commented:
To be complete, you can find the Salaea dump of the communication [here](http://woutergeraedts.nl/dump/salaea-sdmmc-fat-write.logicdata).

thalesfragoso commented:
Hey @Wassasin , thanks for testing.
The first problem was caused because I thought other implementations were going to allocate at least one cluster for a file at creation, but that is not the case, I tested on linux and `touch` only creates the file entry without any cluster allocation. I fixed that in the last commit, now we check if there is a cluster allocated before writing, if not we allocate one. Right now, we allocate one cluster on file creation, should we change that @thejpster ?

About the second issue, unfortunately I can't test with SPI now, but I think I found the problem, it looks like a typo on https://github.com/thejpster/embedded-sdmmc-rs/blob/ef05517d030ea6672fcb6874fc88a91e7c706bab/src/sdmmc.rs#L352

`calc_crc` is `u16`, so I think we should shift it by 8 instead of 16. It's not my code so I'm not sure. CC @thejpster . I also changed that, could you test it again ?

Thanks again for all the testing.

thejpster commented:
> `calc_crc` is `u16`, so I think we should shift it by 8 instead of 16. It's not my code so I'm not sure. CC @thejpster . I also changed that, could you test it again ?

Yes, that's definitely a bug. Without write support for FAT32 I guess the 'write_sector' function hasn't been tested before! Please do change that to '>> 8'. Or use the https://doc.rust-lang.org/std/primitive.u16.html#method.to_be function.

thejpster commented:
Ok, found the other PR and merged it.

thalesfragoso commented:
`README` updated, I also improved the new examples and rebased. If there isn't anything else you would like me to change I think you can merge this, but I would recommend some more testing, especially using SPI (since I couldn't do any) before a crates.io release.

Lucus16 commented:
I ran a basic test using SPI and successfully wrote something and read it back.

thejpster commented:
Alright, let's merge this. Thanks everyone!
